src/common/plist.mligo

```diff
@@ -0,0 +1,142 @@
+          let next =
+            match Map.find_opt (key, true) plist.map with
+              | Some p -> p
+              | None -> ([%Michelson ({| { FAILWITH } |} : (nat * string) -> proposal_key)]
+                  (bad_state, "Expect next link of a key to exist") : proposal_key) in
+          Some { plist with
+            map =
+              (Map.add (next, false) prev // Update the next link
+                (Map.add (prev, true) next // Update the previous link
+                  (Map.remove (key, true) // Remove the next link of the key.
+                    (Map.remove (key, false) plist.map) // Remove the prev link of the key.
+                  )
+                )
+              )
+          }
+        )
+
+
+
+[@inline]
+let plist_update_head (key, plist_o: proposal_key * proposal_doubly_linked_list option): proposal_doubly_linked_list option =
+  match plist_o with
+    | None -> plist_o // Empty plist
+    | Some plist ->
+        if (key = plist.first)
+          then plist_o // Do nothing since the first remain the same
+        else
+          (match Map.find_opt (key, false) plist.map with
+            | Some prev ->
+                (match Map.find_opt (key, true) plist.map with
+                    | None ->
+                        // Previous key exist but no next key, mean the key is the last element.
+                        Some
+                          { first = key
+                          ; last = key
+                          ; map = (Big_map.empty : ((proposal_key * plist_direction), proposal_key) big_map)
+                          }
+                    | Some _ ->
+                        let rec delete_prev (key, plist : proposal_key * proposal_doubly_linked_list): proposal_doubly_linked_list =
+                          match Map.find_opt (key, false) plist.map with
+                            | Some prev ->
+                                (match Map.find_opt (key, true) plist.map with
+                                  | Some _ ->
+                                      let new_plist = {plist with
+                                            map = Map.remove (key, true)
+                                                    (Map.remove (key, false) plist.map)
+                                          } in
+                                      delete_prev (prev, new_plist)
+                                  | None -> plist // Shouldn't happen
+                                )
+                            | None -> plist in
+
+                        let new_plist = delete_prev (prev, plist) in
+
+                        Some { new_plist with first = key }
+
+                )
+            | None ->
+                ([%Michelson ({| { FAILWITH } |} : (nat * string) -> (proposal_doubly_linked_list option))]
+                    (bad_state, "Error updating head with a key that does not exist.") : (proposal_doubly_linked_list option))
+          )
+
+
+#endif // COMMON_PLIST_H included
```

Comment:
It seems that we have quite a bit of logic here.

Can we add hedgehog property tests to ensure the soundness of logic here (May be by comparing the behviour to data structures from `ordered-containers` package, since this seem to implement the same functionality?). 

I think we should also have separate SMT tests to target this section, as the subset of main SMT tests might not provide enough variance to properly test this section.

Reply:
Make sense, I'll try to work on this and see if it could work out.